### DISCUSSION TITLE
Triggering database migrations in serverless world

### DISCUSSION THREAD

**Discussion by kennethacohen** [vercel/next.js#15650] [Help]
Realize this isn't directly related to Next, but facing some new problems in a shift to API routes and curious how other folks manage.

What is the correct pattern for deploying new code + triggering database migrations alongside deploys? This made sense to me in the Express world (given a persistent db connection), but not sure what the recommended path is for serverless.

**Comment by samcx** [comment]
Hello,

Serverless functions are not persistent and not always "on" like a Node.js Express server. API routes are great for use cases such as handling user authentication, form submission, database queries, custom slack commands, and more -> https://vercel.com/docs/v2/serverless-functions/introduction#.

**Reply by kennethacohen** [reply]
Yep, I have loads of Next API routes built out..so it seems like I do need a persistent server in addition to these routes? i.e. I should operate in a mixed server world vs going full serverless..is that a common pattern?

**Reply by DeBraid** [reply]
I have similar issues.  Previous stack used Node Express server, where I could trigger database migrations by executing `node` scripts on the server.  No longer the case with new NextJS stack it appears. 

As potential hack workaround, I've considered creating a new API route which I can arbitrarily hit w/ `curl`.  To ensure this only gets executed once, the API route would require a token.  Something like `/api/users/update-roles?token=${SECRETS}`.

Have not built any code around this concept, still exploring options.  @kennethacohen mind sharing how you solved this problem? Thanks.

**Reply by kennethacohen** [reply]
To be honest I solved it by...not using serverless. Couldn't find enough docs/support to answer various questions, so just went back to the old world :/

**Reply by AHBruns** [reply]
I've been doing database migrations as part of my builds. This is potentially unsafe if the database change isn't compatible with the current application code. So, I use tick-tock release scheduling where each build has to be either a database schema change w/ no source code changes or a source code change with no database schema changes. This is all handled by a GitHub workflow.

**Comment by DeBraid** [comment]
Reproduced from: https://github.com/vercel/next.js/discussions/17479#discussioncomment-434551

TIL, you can run nodejs scripts at build time with `webpack` config.  Below, in the context of sitemap generation:

> Finally, override your `next.config.js` to run this script as part of the build process. Your generated file gets created at public/sitemap.xml which then gets moved to the root of your site (requires Next v9.0+).

```
module.exports = {
  webpack: (config, { isServer }) => {
    if (isServer) {
      require('./scripts/generate-sitemap');
    }

    return config;
  }
};
```

via https://leerob.io/blog/nextjs-sitemap-robots

**Comment by marcelovicentegc** [comment]
I hope this helps those who come across this. This is something I do no matter the environment (be it serverless or not), and it has worked pretty well! 

On each new deployment, the `build` command is executed and migrations run. 

package.json
```json
"scripts": {
    "dev": "yarn db:migrate && next dev",
    "build": "yarn db:migrate && next build",
    "start": "next start",
    "db:migrate": "ts-node --project ./tsconfig.database.json -r dotenv/config ./node_modules/typeorm/cli.js migration:run",
    "db:makemigrations": "ts-node --project ./tsconfig.database.json -r dotenv/config ./node_modules/typeorm/cli.js migration:generate --name",
    "db:undomigration": "ts-node --project ./tsconfig.database.json -r dotenv/config ./node_modules/typeorm/cli.js migration:revert"
  },
```

cc @kennethacohen

**Reply by GabenGar** [reply]
Does it mean `ts-node` has to be a production dependency?

**Reply by richardscarrott** [reply]
@GabenGar I _think_ it would have to be yeah...but then again @marcelovicentegc appears to only be running `.js` with it so could prob switch it out for `node` here?

If you are using TS, perhaps it's better to use `tsc` to compile your `.ts` files into `.js`, then you wouldn't need TS as a production dep, but of course special nextjs features such as importing css files wouldn't work.

**Reply by manuganji** [reply]
Maybe it's more ideal to wait for the migrate happen after a successful build? How about 

`"build": "next build && yarn db:migrate",`

**Comment by pthieu** [comment]
@marcelovicentegc how do you define the env var for your database url? this would only work if you set it in your CI/CD env and I'm trying to avoid that

**Reply by marcelovicentegc** [reply]
Yes, I define it on the CI/CD pipeline, yes! What's your concern with setting the env variables on the CI/CD pipeline?

**Reply by pthieu** [reply]
I generally don't like giving CI/CD more perms than needed, but migrations is acceptable, just wanted to see if I could avoid it.

**Comment by mytabworks** [comment]
I wonder how do you run a migration script and read the .env file when the declared Environment Variable are save on Vercel, not on your local source code?

**Reply by GabenGar** [reply]
Vercel env vars are available with the same `process.env.*` call as the source code. You just have to set the variables in its own dashboard beforehand.
